### PR TITLE
Update XGrammar to the latest API

### PR DESCRIPTION
**Pull request by Ubospica**
## Motivation
After XGrammar released their 0.10 update, there was a good bit of refactoring which seems to have broken the initial integration with SGLang. This PR works to resolve the issues and make XGrammar a working backend for SGLang.

Thanks to the great work by @gittb.

Co-authored-by: Ben Gitter <gitterbd@gmail.com>

### CODE REVIEW

**Review by zhyncs — APPROVED** [review]

**Comment by Ubospica** [comment]
Need to wait for #2175 to get merged first

**Comment by zhyncs** [comment]
BTW may you update the xgrammar version in pyproject.yml

**Comment by Ubospica** [comment]
> BTW may you update the xgrammar version in pyproject.yml

Sounds good

File: test/srt/test_json_constrained.py [hunk 1]
```diff
@@ -121,5 +121,114 @@ def test_mix_json_and_other(self):
             list(executor.map(self.run_decode, json_schemas))
 
 
+class TestJSONConstrainedXGrammarBackend(unittest.TestCase):
```
Comment (merrymercy) [inline test/srt/test_json_constrained.py]: Do not copy the code. You can inherit the above class and only rewrite the setUpClass function.

**Review by merrymercy — CHANGES_REQUESTED** [review]

**Comment by zhyncs** [comment]
We also need to install the dependencies. ref https://xgrammar.mlc.ai/docs/start/install#id3
```bash
pip3 install cuda-python nvidia-cuda-nvrtc-cu12
```

**Comment by zhyncs** [comment]
```bash
# v0.3.6
docker pull lmsysorg/sglang:latest
docker run -itd --gpus all --shm-size 32g 57294f0977f9  /bin/bash
docker exec -it id /bin/bash
mkdir tmp && cd tmp
git clone -b main-dev/2024-11-25-update-xgrammar --depth=1 https://github.com/Ubospica/sglang
pip3 install -e python[all] --force-reinstall
pip3 install cuda-python nvidia-cuda-nvrtc-cu12
cd test/srt
python3 test_json_constrained.py
```

```
[2024-11-25 07:54:08 TP0] Use CUDA home /usr/local/cuda
sourceCode.cu(13): error: identifier "int32_t" is undefined
      const int32_t* __restrict__ bitmask,
            ^

sourceCode.cu(14): error: identifier "int32_t" is undefined
      const int32_t* __restrict__ indices,
            ^

sourceCode.cu(25): error: identifier "int32_t" is undefined
      if (GET_BIT(reinterpret_cast<const int32_t*>(bitmask + blockIdx.y * bitmask_size), tid + i) ==
          ^

3 errors detected in the compilation of "sourceCode.cu".

CUDA error code=6(b'NVRTC_ERROR_COMPILATION')
Exception in thread Thread-4 (forward_thread_func):
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/xgrammar/kernels/apply_token_bitmask_inplace_cuda.py", line 170, in compile
    checkCudaErrors(nvrtc.nvrtcCompileProgram(prog, len(opts), opts))
  File "/usr/local/lib/python3.10/dist-packages/xgrammar/kernels/apply_token_bitmask_inplace_cuda.py", line 88, in checkCudaErrors
    raise RuntimeError(
RuntimeError: CUDA error code=6(b'NVRTC_ERROR_COMPILATION')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/sgl-workspace/tmp/sglang/python/sglang/srt/managers/tp_worker_overlap_thread.py", line 91, in forward_thread_func
    self.forward_thread_func_()
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/sgl-workspace/tmp/sglang/python/sglang/srt/managers/tp_worker_overlap_thread.py", line 118, in forward_thread_func_
    logits_output, next_token_ids = self.worker.forward_batch_generation(
  File "/sgl-workspace/tmp/sglang/python/sglang/srt/managers/tp_worker.py", line 149, in forward_batch_generation
    next_token_ids = self.model_runner.sample(logits_output, model_worker_batch)
  File "/sgl-workspace/tmp/sglang/python/sglang/srt/model_executor/model_runner.py", line 652, in sample
    logits = self.apply_logits_bias(logits_output.next_token_logits, sampling_info)
  File "/sgl-workspace/tmp/sglang/python/sglang/srt/model_executor/model_runner.py", line 677, in apply_logits_bias
    sampling_info.apply_mask(logits=logits, vocab_mask=sampling_info.vocab_mask)
  File "/sgl-workspace/tmp/sglang/python/sglang/srt/constrained/xgrammar_backend.py", line 94, in apply_vocab_mask
    apply_token_bitmask_inplace(logits, vocab_mask)
  File "/usr/local/lib/python3.10/dist-packages/xgrammar/matcher.py", line 110, in apply_token_bitmask_inplace
    apply_token_bitmask_inplace_cuda(logits, bitmask, indices)
  File "/usr/local/lib/python3.10/dist-packages/xgrammar/kernels/apply_token_bitmask_inplace_cuda.py", line 235, in apply_token_bitmask_inplace_cuda
    kernel = KernelStore.compile(logits.device.index)
  File "/usr/local/lib/python3.10/dist-packages/xgrammar/kernels/apply_token_bitmask_inplace_cuda.py", line 178, in compile
    raise RuntimeError("CUDA kernel compilation failure")
RuntimeError: CUDA kernel compilation failure
```

**Pull request merged**